subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ProviderInternal.java

```diff
@@ -24,6 +24,88 @@
 import javax.annotation.Nullable;
 import java.util.function.BiFunction;
 
+/**
+ * <p>The internal view of a {@link Provider}.</p>
+ *
+ * <p>A {@code Provider<T>} has several different pieces of state associated with it:</p>
+ *
+ * <ul>
+ *     <li>The "value" of the provider, which is the object returned by {@link #get()} and its variants.</li>
+ *     <li>the "content" of the value. The value points to some object, but this object may not be immutable even though the value is fixed.
+ *     For example, when the value is a {@code File} object, the state of that file in the file system may change. This mutable state is the "content" of the value.</li>
+ * </ul>
+ *
+ * <p>In other words, a provider does not provide a value, it provides a value whose content is in a particular state. This is discussed in more detail below.</p>
+ *
+ * <h1>The provider value</h1>
+ *
+ * The value of a provider may be:</p>
+ *
+ * <ul>
+ *     <li>"fixed", which means that the value is a function only of configuration time inputs.
+ *     For example, when the value of a task property is defined using a constant in a build script, then that property (which is-a provider) has a "fixed" value and its inputs include the build script.
+ *     <p>When a provider has a fixed value, then the value of the provider can be assumed to have a constant ("fixed") value during execution time.
+ *     In particular, this means that the fixed value can be written to the configuration cache entry rather than writing the provider itself, which is good for performance.</p>
+ *     <p>There are other ways this concept could be used but that are not yet implemented,
+ *     for example a provider with a fixed value could potentially memoize its value when queried at execution time, rather than calculating the value each time it is queried.</p>
+ *     </li>
+ *     <li>"changing", which means that the value is a function of some execution time inputs.
+ *     For example, a task output property (whose value is a function of the task inputs) typically has a changing value.
+ *     <p>When a provider has a changing value, then the provider must be written to the configuration cache, rather than writing its current value. This process is optimized by replacing any
+ *     input providers that have a fixed value with that value, as above.
+ *     </p>
+ *     </li>
+ * </ul>
+ *
+ * <p>See {@link org.gradle.api.internal.provider.ValueSupplier.ExecutionTimeValue}, which represents these states.</p>
+ *
+ * <p>The value itself might be "missing", which means there is no value available, or "broken", which means the calculation failed with some exception, or some object.</p>
```

Comment:
The "broken" value is mentioned here, but it is not clear how to understand if the value is broken from the APIs that we have, since there is no flag that we store for this. Could you elaborate on that?

Do we have any docs on how Providers should generally behave in case of failures?